Reduce to single Exec, improve validation

reidmv opened the pull request:
Probably easiest to just review the [end result](https://github.com/reidmv/file_permissions/blob/improvements/manifests/init.pp).

Pros:
Performance. Single exec resource.

Cons:
More complicated Puppet code.

reidmv committed:
Refactor: use only one Exec resource

Separate logic of figuring out what command to run from resource to run
command. This is for efficiency and clarity when running.

npwalker commented:
@reidmv you failed linting.

```
manifests/init.pp - WARNING: string containing only a variable on line 52
manifests/init.pp - WARNING: quoted boolean value found on line 73
``` 

and it looks like you need to rebase

reidmv commented:
Rebased, lint issues fixed.

npwalker commented on .fixtures.yml:
```diff
@@ -3,4 +3,4 @@
 fixtures:
   forge_modules:
-#     stdlib: "puppetlabs/stdlib"
+    stdlib: "puppetlabs/stdlib"
```
We opted to skip the stdlib dependency and just copy the types over.

reidmv replied:
This implementation uses `join()`, so has a dependency on either stdlib or on Puppet 5.5. Given that it has a dependency on stdlib anyway I figured we might as well use the standard types rather than copying them.

If you prefer, I can rework this to express a dependency on an older stdlib. IMO that's not as clean a solution though.

npwalker replied:
IMO the dependency on too new of a stdlib is going to prevent use of the module.  So, using reduce instead of join avoids the need for stdlib.

npwalker commented on manifests/init.pp:
```diff
@@ -28,38 +27,64 @@
+  # deviation from the desired state of any validator.
+  $command = $validators.reduce('true') |$memo,$validator| { # lint:ignore:quoted_booleans
+    if !$validator[input].empty() {
+      [$memo, "find ${shellsafe_dir} '(' ${validator[find]} ')' ${validator[fix]}"].join(' && ')
```
a poor man's join. 

```
.reduce |$result, $value|  { "${result} && ${value}" }
```

reidmv commented:
@npwalker `reduce()` for `join()` eh? Well, that's clever. Updated to remove stdlib dependency.

reidmv committed:
Update Puppet dependency to >=4.0.0

Because we're using the reduce() function, and types.

reidmv committed:
Add support for OSX

On OSX, chmod and friends do not support the -c flag. They do support
-v, which just prints the names of files the command operates on.

For OSX, use -v. Better than not working. Not quite as beautiful in the
output.

natemccurdy commented:
This looks great!

But what about Solaris? Seems like its `chmod` and `chown` command don't support `-v` or `-c` based on some [random Googling](https://docs.oracle.com/cd/E19109-01/tsolaris8/817-0879/6mgl9vnf9/index.html).

Anyone got a box to test that on?

reidmv commented:
Confirmed that -v doesn't work on AIX. I'm happy to assume that applies to Solaris also.

reidmv committed:
Add AIX and Solaris support

They have versions of ch{own,mod,grp} that don't support cool flags.
Accomodate them anyway!

npwalker commented:
+1 to premature optimization
